src/core/protocol-authorization.ts

```diff
@@ -84,10 +102,31 @@ export class ProtocolAuthorization {
    * Constructs a chain of ancestor messages
    * @returns the ancestor chain of messages where the first element is the root of the chain; returns empty array if no parent is specified.
    */
-  private static async constructAncestorMessageChain(tenant: string, recordsWrite: RecordsWrite, messageStore: MessageStore)
+  private static async constructAncestorMessageChain(
+    tenant: string,
+    incomingMessage: RecordsRead | RecordsWrite,
+    messageStore: MessageStore
+  )
     : Promise<RecordsWriteMessage[]> {
     const ancestorMessageChain: RecordsWriteMessage[] = [];
 
+    // Get first RecordsWrite in ancestor chain, or use incoming write message
+    let recordsWrite: RecordsWrite;
+    if (incomingMessage.message.descriptor.method === DwnMethodName.Write) {
+      recordsWrite = incomingMessage as RecordsWrite;
+    } else {
+      const recordsRead = incomingMessage as RecordsRead;
+      const query = {
+        interface : DwnInterfaceName.Records,
+        method    : DwnMethodName.Write,
+        recordId  : recordsRead.message.descriptor.recordId,
+      };
+      const existingMessages = await messageStore.query(tenant, query) as TimestampedMessage[];
```

Comment:
Seeing this, I can't help to think that this is a faster and simpler universal approach to do the chain construction regardless of method type? ie. one query that fetches all messages of the given recordId then construct the chain. What do you think?

Reply:
To get all the ancestor messages, we have to query for different records with different recordIds, since each ancestor is a separate `RecordsWrite`. We need to successively query by a different `parentId` for each message in the chain.

It's possible I'm misunderstanding your suggestion, but I don't think we can reduce down to one query by recordId.

Reply:
only way i can think of to use a single query would be to query using `contextId` _only_. then you'd have to search the result-set for the record that has a `recordId === parentId` of incoming message and start building the ancestor chain from there. and i think you'd still have to query for the newest message for any given `recordId` i believe.

there may be more optimal ways that i'm sure we can get into later. for now i feel like this is very easy to read/follow which really helps think through whether it's correct

Reply:
@diehuxx i'm wondering if we need to check whether a given record in the ancestral chain has been deleted?

Reply:
Brings up some good questions, namely: what happens to records that descend from deletes? I think there are a couple ways to look at it:

1. It should cause all descendants to be deleted.
2. Because of tombstoning, the chain is able to be retained. I'd personally favor #1, because app use cases generally expect it, and it's hard to reason about things if we don't keep it simple.

Reply:
@csuwildcat Deleting all descendants seems too restrictive. For example, in a chat app with threads, if someone deletes on message in the thread, I wouldn't expect all subsequent messages to disappear. I know that Slack works this way: if one chat message disappears, the deleted message disappears or is replaced with "Message Deleted" and the rest of the thread is untouched. I've definitely seen other chat apps operate similarly.

Now I'm gonna devils advocate myself. The "Message Deleted" UI might be supportable by sending a `RecordsWrite` that overwrites the data to just say "Message Deleted" or something. In that case, the app developer could prohibit `RecordsDelete` for such records, so they wouldn't need to care about whether descendants get deleted.

So, it's possible that clever app developers use workarounds like this to "delete" single messages in the chain. But it still feels too restrictive, and IMO hard to reason about if descendants are always deleted.

Just thinking out loud: Protocol authz'd deletes aren't implemented yet. What if we had both `delete` and `delete-descendants` actions in protocols. Give developers both options.

Reply:
@diehuxx and i chatted and we both feel like addressing this issue is out of scope for this PR. @diehuxx is going to create an issue for the topic brought up in this comment and we'll address it separately

Reply:
Moving discussion to a separate issue as record deletion is out of scope for this PR
#311